lib/api/CDataFrameAnalyzer.cc

```diff
@@ -6,15 +6,133 @@
+// Control message types:
+const char RUN_ANALYSIS_CONTROL_MESSAGE_FIELD_VALUE{'r'};
+}
+
+// TODO memory calculations plus choose data frame storage strategy.
+CDataFrameAnalyzer::CDataFrameAnalyzer(CDataFrameAnalysisSpecification analysisSpecification)
+    : m_AnalysisSpecification{std::move(analysisSpecification)},
+      m_DataFrame{core::makeMainStorageDataFrame(m_AnalysisSpecification.cols())} {
+}
+
+bool CDataFrameAnalyzer::usingControlMessages() const {
+    return m_ControlFieldValue >= 0;
+}
 
 bool CDataFrameAnalyzer::handleRecord(const TStrVec& fieldNames, const TStrVec& fieldValues) {
-    LOG_INFO(<< core::CContainerPrinter::print(fieldNames));
-    LOG_INFO(<< core::CContainerPrinter::print(fieldValues));
+
+    // Control messages are signified by a dot in the field name. This supports:
+    //   - using the last field for a control message,
+    //   - missing.
+    //
+    // Note if the the control message field name is missing the analysis must
+    // be triggered to run by calling run explicitly.
+
+    bool result{true};
+    if (this->readyToReceiveControlMessages() == false) {
+        result = this->prepareToReceiveControlMessages(fieldNames);
+    }
+
+    if (this->isControlMessage(fieldValues)) {
+        return this->handleControlMessage(fieldValues);
+    }
+
+    this->addRowToDataFrame(fieldValues);
+
+    return result;
+}
+
+void CDataFrameAnalyzer::run() {
+    // TODO
+}
+
+bool CDataFrameAnalyzer::readyToReceiveControlMessages() const {
+    return m_ControlFieldValue != CONTROL_FIELD_UNSET;
+}
+
+bool CDataFrameAnalyzer::prepareToReceiveControlMessages(const TStrVec& fieldNames) {
+    if (fieldNames.empty() || fieldNames.back() != CONTROL_MESSAGE_FIELD_NAME) {
+        m_ControlFieldValue = CONTROL_FIELD_MISSING;
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = static_cast<std::ptrdiff_t>(fieldNames.size());
+    } else {
+        m_ControlFieldValue = static_cast<std::ptrdiff_t>(fieldNames.size() - 1);
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = m_ControlFieldValue;
+        auto pos = std::find(fieldNames.begin(), fieldNames.end(), CONTROL_MESSAGE_FIELD_NAME);
+        if (pos != fieldNames.end()) {
+            LOG_ERROR(<< "Unexpected possible control field: ignoring");
+            return false;
+        }
+    }
     return true;
 }
+
+bool CDataFrameAnalyzer::isControlMessage(const TStrVec& fieldValues) const {
+    return m_ControlFieldValue != CONTROL_FIELD_MISSING &&
+           fieldValues[m_ControlFieldValue].size() > 0;
+}
+
+bool CDataFrameAnalyzer::handleControlMessage(const TStrVec& fieldValues) {
+    bool unrecognised{fieldValues.size() > 1};
+    if (unrecognised == false) {
+        switch (fieldValues[m_ControlFieldValue][0]) {
+        case RUN_ANALYSIS_CONTROL_MESSAGE_FIELD_VALUE:
+            this->run();
+            break;
+        default:
+            unrecognised = true;
+        }
```

Comment:
I think it's nice to have a `break` for the `default` case too, even though it won't change the control flow in this instance.